src/providers/aws.rs

```diff
@@ -727,41 +735,69 @@ impl RegionLauncher {
     /// Make spot instance requests, wait for the instances, and then call the
     /// instance setup functions.
     #[instrument(level = "debug", skip(self, max_wait))]
-    pub async fn launch<M>(
-        &mut self,
-        mode: LaunchMode,
+    pub async fn launch<'l, M>(
+        &'l mut self,
+        mut mode: LaunchMode,
         mut max_wait: Option<time::Duration>,
         machines: M,
     ) -> Result<(), Report>
     where
         M: IntoIterator<Item = (String, Setup)> + std::fmt::Debug,
     {
-        match mode {
-            LaunchMode::DefinedDuration {
-                hours: max_instance_duration_hours,
-            } => {
-                self.make_spot_instance_requests(
-                    max_instance_duration_hours * 60, // 60 mins/hr
-                    machines,
-                )
-                .await
-                .wrap_err("failed to make spot instance requests")?;
+        let machines: Vec<_> = machines.into_iter().collect();
 
-                let start = time::Instant::now();
-                self.wait_for_spot_instance_requests(max_wait)
-                    .await
-                    .wrap_err("failed while waiting for spot instances fulfilment")?;
-                if let Some(ref mut d) = max_wait {
-                    *d -= time::Instant::now().duration_since(start);
+        loop {
```

Comment:
I don't think this needs to be a `loop`. Just check spot first, and early-return if it succeeds?

Reply:
With the loop, we end up not having to copy-paste the call to `wait_for_instances` twice. This also relates to why there was a lifetime hanging around from my first attempt at this - I tried a way with a closure with early return instead of the loop, but ended up deciding the result was too complicated.
Another way is to have a helper function that does this part, if you think that would be more readable.